packages/@ember/application/lib/application.js

```diff
@@ -483,13 +483,22 @@ const Application = Engine.extend({
     @method waitForDOMReady
   */
   waitForDOMReady() {
-    if (!this.$ || this.$.isReady) {
+    if (!hasDOM || this.documentReadyState() !== 'loading') {
```

Comment:
IMO we can just inline `document.readyState` here. As you pointed out, `hasDOM` already checked that `document` exists so you don't need to check it again.

Reply:
I had to add the `documentReadyState()` method to be able to test it. Couldn't find another way to mock a read-only `readyState` property 🥺

Reply:
In fact, I didn't find a way to restore `readyState` after `Object.defineProperty(document, "readyState", ...)`, so it was polluting the global state.

Reply:
Yeah, that's a good point, but in that case we have bigger problems, e.g. we are leaking event handlers on the actual global document object in tests. I'm not sure what's the best approach here, but if we are going to mock, we should just mock the whole document object (the parts that we need), including `addEventListener` and `removeEventListener`. I'll probably pass it in to the constructor/init of `Application` and default it to the global document object.